### PR TITLE
[4.3] Migrate most components to the new toolbar api

### PR DESCRIPTION
**Pull request by wilsonge**
### Summary of Changes
Migrate the majority of core components towards the new toolbar API (notably I'm skipping Joomla Update as part of this PR and the use of ToolbarHelper::modal in com_templates views which probably need to be moved to the popupButton - but that's a bigger piece of work for another PR)

### Testing Instructions
Check all the toolbar icons work the same before and after the patch is applied across all backend components

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Review by wojtekxtx — APPROVED** [review]
LGTM :)

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on b4bab7d8c64aa3d421a2d27f8efd3bf22b64a760<br /><br />changes the button id :(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39537">issues.joomla.org/tracker/joomla-cms/39537</a>.</sub>

**Comment by wilsonge** [comment]
https://github.com/joomla/joomla-cms/blob/4.2-dev/libraries/src/Toolbar/ToolbarButton.php#L221

IDs look fairly fixed. I can rebase this to 5.0 I guess - Depends how much we think IDs are a major b/c break in the backend. It's only the delete button that has the changed ID right?

**Comment by wojtekxtx** [comment]
> > changes the button id :(

@brianteeman: it should not happen, but, its not huge change. Im ok with it.

**Review by wojtekxtx — APPROVED** [review]
LGTM :)

**Comment by carlitorweb** [comment]
@wilsonge what you mean with **cache toolbar** in the test instructions? Is just test the toolbar of each component right? Sorry for ask, just testing this PR, and for avoid do it wrong.

**Comment by wilsonge** [comment]
Yes. Sorry originally the PR was just over com_cache but then I worked on some more components so need to update the instructions

**Comment by wilsonge** [comment]
Final batch of components migrated. I suggest we test functionality here and I'll leave it to @joomla/cms-maintainers to decide if it needs to go to 4.3 or 5.0 and I'll happily rebase it to 5.0 if the IDs are deemed a concern.

**Review by wojtekxtx — APPROVED** [review]
LGTM :)

Pretty much its ready.

**Comment by carlitorweb** [comment]
My progress so far in the testing (will test the rest in some hours)

- com_actionlogs ✔️ 
- com_admin ✔️
- com_associations ⏳ 
  - banner ✔️
  - banners ✔️
  - client ✔️
  - clients ✔️
  - tracks ⏳
- com_cache ✔️
- com_categories ✔️
- com_checkin ✔️
- com_config ✔️
- com_contact ✔️
- com_cpanel ✔️
- com_content ✔️
- com_contenthistory ✔️
- com_fields ⏳
- com_finder ⏳
- com_installer ⏳
- com_languages ⏳
- com_media ⏳
- com_menus ⏳
- com_messages ⏳
- com_modules ⏳
- com_newsfeed ⏳
- com_plugins ⏳
- com_postinstall ⏳
- com_privacy ⏳
- com_redirect ⏳
- com_sheduler ⏳
- com_tags ⏳
- com_templates ⏳
- com_users ⏳
- com_workflow ⏳
- com_wokflow ⏳

**Comment by carlitorweb** [comment]
@wilsonge com_installer -> view=manage

The button **Install Extension**: Undefined variable $url in ..www\bugtesting\joomla-cms\layouts\joomla\toolbar\link.php on line 34

**Comment by wojtekxtx** [comment]
@carlitorweb also got this error

**Comment by wilsonge** [comment]
Fixed!

File: administrator/components/com_associations/src/View/Association/HtmlView.php [hunk 1]
```diff
@@ -359,29 +359,24 @@ protected function addToolbar(): void
...
             . 'class="btn btn-success"><span class="icon-save" aria-hidden="true"></span>'
-            . Text::_('COM_ASSOCIATIONS_SAVE_TARGET') . '</button></joomla-toolbar-button>',
-            'target'
-        );
+            . Text::_('COM_ASSOCIATIONS_SAVE_TARGET') . '</button></joomla-toolbar-button>');
 
         if ($this->typeName === 'category' || $this->extensionName === 'com_menus' || $this->save2copy === true) {
-            ToolbarHelper::custom('copy', 'copy.png', '', 'COM_ASSOCIATIONS_COPY_REFERENCE', false);
+            $toolbar->standardButton('', 'COM_ASSOCIATIONS_COPY_REFERENCE', 'copy')
+                ->icon('copy.png')
```
Comment (Quy) [inline administrator/components/com_associations/src/View/Association/HtmlView.php]:
```suggestion
                ->icon('icon-copy')
```

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 419835624b7c2c755b7781606f4935334a541e77<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39537">issues.joomla.org/tracker/joomla-cms/39537</a>.</sub>

File: administrator/components/com_contact/src/View/Contacts/HtmlView.php [hunk 2]
```diff
@@ -165,9 +166,7 @@ protected function addToolbar()
             $childBar = $dropdown->getChildToolbar();
 
             $childBar->publish('contacts.publish')->listCheck(true);
-
             $childBar->unpublish('contacts.unpublish')->listCheck(true);
-
             $childBar->standardButton('featured')
                 ->text('JFEATURE')
                 ->task('contacts.featured')
```
Comment (Quy) [inline administrator/components/com_contact/src/View/Contacts/HtmlView.php]:
```suggestion
            $childBar->standardButton('featured', 'JFEATURE', 'contacts.featured')
```
Please do the following line as it is not available in the GitHub GUI.

Comment (wojtekxtx) [reply administrator/components/com_contact/src/View/Contacts/HtmlView.php:172]: @Quy what?

File: administrator/components/com_config/src/View/Component/HtmlView.php [hunk 3]
```diff
@@ -109,19 +110,21 @@ public function display($tpl = null)
...
-        ToolbarHelper::divider();
-        ToolbarHelper::save('component.save');
-        ToolbarHelper::divider();
-        ToolbarHelper::cancel('component.cancel', 'JTOOLBAR_CLOSE');
-        ToolbarHelper::divider();
+        $toolbar->apply('component.apply');
+        $toolbar->divider();
+        $toolbar->save('component.save');
+        $toolbar->divider();
+        $toolbar->cancel('component.cancel', 'JTOOLBAR_CLOSE');
```
Comment (Quy) [inline administrator/components/com_config/src/View/Component/HtmlView.php]:
```suggestion
        $toolbar->cancel('component.cancel');
```

File: administrator/components/com_categories/src/View/Category/HtmlView.php [hunk 4]
```diff
@@ -189,61 +191,66 @@ protected function addToolbar()
...
-                'btn-success'
+                    // If an existing item, can save to a copy.
+                    if ($canDo->get('core.create')) {
+                        $childBar->save2copy('category.save2copy');
+                    }
+                }
             );
 
-            ToolbarHelper::cancel('category.cancel', 'JTOOLBAR_CLOSE');
+            $toolbar->cancel('category.cancel', 'JTOOLBAR_CLOSE');
```
Comment (Quy) [inline administrator/components/com_categories/src/View/Category/HtmlView.php]:
```suggestion
            $toolbar->cancel('category.cancel');
```

File: administrator/components/com_finder/src/View/Filter/HtmlView.php [hunk 5]
```diff
@@ -128,48 +130,49 @@ protected function addToolbar()
...
+
+                    // If an existing item, can save as a copy
+                    if ($canDo->get('core.create')) {
+                        $childBar->save2copy('filter.save2copy');
+                    }
+                }
             );
 
-            ToolbarHelper::cancel('filter.cancel', 'JTOOLBAR_CLOSE');
+            $toolbar->cancel('filter.cancel', 'JTOOLBAR_CLOSE');
```
Comment (Quy) [inline administrator/components/com_finder/src/View/Filter/HtmlView.php]:
```suggestion
            $toolbar->cancel('filter.cancel');
```

File: administrator/components/com_templates/src/View/Template/HtmlView.php [hunk 6]
```diff
@@ -322,13 +323,13 @@ protected function addToolbar()
         }
 
         if ($this->type === 'home') {
-            ToolbarHelper::cancel('template.cancel', 'JTOOLBAR_CLOSE');
+            $toolbar->cancel('template.cancel', 'JTOOLBAR_CLOSE');
```
Comment (Quy) [inline administrator/components/com_templates/src/View/Template/HtmlView.php]:
```suggestion
            $toolbar->cancel('template.cancel');
```

File: administrator/components/com_content/src/View/Featured/HtmlView.php [hunk 7]
```diff
@@ -183,13 +180,11 @@ protected function addToolbar()
...
 
-                $childBar->standardButton('unfeatured')
-                    ->text('JUNFEATURE')
+                $childBar->standardButton('unfeatured', 'JUNFEATURE')
                     ->task('articles.unfeatured')
                     ->listCheck(true);
 
                 $childBar->archive('articles.archive')->listCheck(true);
-
                 $childBar->checkin('articles.checkin')->listCheck(true);
```
Comment (Quy) [inline administrator/components/com_content/src/View/Featured/HtmlView.php]:
```suggestion
                $childBar->checkin('articles.checkin');
```

File: administrator/components/com_contact/src/View/Contacts/HtmlView.php (hunk 2 above)
Comment (obuisard) [reply administrator/components/com_contact/src/View/Contacts/HtmlView.php]: both are valid, second makes it shorter

File: administrator/components/com_content/src/View/Featured/HtmlView.php (hunk 7 above)
Comment (carlitorweb) [reply administrator/components/com_content/src/View/Featured/HtmlView.php]: In this case should thrown a error right? I did not get that in the test

**Comment by HLeithner** [comment]
Is there a reason why you not directly migrate this to `ListView` and `FormView`, also `Toolbar::getInstance();` is deprecated so I think you are not allowed to introduce new deprecated function calls ;-)

**Comment by obuisard** [comment]
I have tested this item :white_check_mark: successfully on fd47b40fcdb13ca65c69c9e7631b06d9a65e489b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39537">issues.joomla.org/tracker/joomla-cms/39537</a>.</sub>

File: administrator/components/com_contact/src/View/Contacts/HtmlView.php [hunk 8]
```diff
@@ -165,22 +166,17 @@ protected function addToolbar()
...
             $childBar->unpublish('contacts.unpublish')->listCheck(true);
-
-            $childBar->standardButton('featured')
-                ->text('JFEATURE')
-                ->task('contacts.featured')
+            $childBar->standardButton('featured', 'JFEATURE', 'contacts.featured')
                 ->listCheck(true);
             $childBar->standardButton('unfeatured')
                 ->text('JUNFEATURE')
                 ->task('contacts.unfeatured')
```
Comment (Quy) [inline administrator/components/com_contact/src/View/Contacts/HtmlView.php]:
```suggestion
            $childBar->standardButton('unfeatured', 'JUNFEATURE', 'contacts.unfeatured')
```

**Comment by wilsonge** [comment]
> Is there a reason why you not directly migrate this to ListView and FormView, also Toolbar::getInstance(); is deprecated so I think you are not allowed to introduce new deprecated function calls ;-)

Not all of these views are ready to be migrated - but having everything consistently using the toolbar api as an example to extension devs is a good thing.

Toolbar::getInstance removal is a major deal. The factory being stored in the container means that there's no way to share it with the module. I suspect that's a major version change.

**Comment by bembelimen** [comment]
> Not all of these views are ready to be migrated 

But shouldn't we move the one we can to the two mentioned views (+ updating them, too): It would save so much code.

File: administrator/components/com_content/src/View/Featured/HtmlView.php (hunk 7 above)
Comment (Quy) [reply administrator/components/com_content/src/View/Featured/HtmlView.php]: `listCheck(true)` has been added to:

```
    public function checkin(string $task, string $text = 'JTOOLBAR_CHECKIN'): StandardButton
    {
        return $this->standardButton('checkin', $text, $task)
            ->listCheck(true);
    }
```

**Comment by wilsonge** [comment]
> But shouldn't we move the one we can to the two mentioned views (+ updating them, too): It would save so much code.

Eventually yes. But I don't think we are ready yet. Certainly aligning the views first so that jump is smaller isn't a bad thing anyhow

File: administrator/components/com_contact/src/View/Contacts/HtmlView.php (hunk 2 above)
Comment (wilsonge) [reply administrator/components/com_contact/src/View/Contacts/HtmlView.php]: It's not just about shorter. the way most IDE's interpret the return types mean that both linters and IDE's interpret stuff better.

File: libraries/src/Toolbar/CoreButtonsTrait.php [hunk 9]
```diff
@@ -125,6 +126,22 @@ public function help($ref, $useComponent = false, $url = null, $component = null
...
+    /**
+     * Writes a help button for a given option (opens a popup window).
+     *
+     * @param   string  $class   The class used by the inline help items.
+     *
+     * @return  InlinehelpButton
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function inlinehelp(string $class = "hide-aware-inline-help"): InlinehelpButton
```
Comment (Quy) [inline libraries/src/Toolbar/CoreButtonsTrait.php]:
```suggestion
    public function inlinehelp(string $class = 'hide-aware-inline-help'): InlinehelpButton
```

File: administrator/components/com_redirect/src/View/Links/HtmlView.php [hunk 10]
```diff
@@ -193,22 +193,19 @@ protected function addToolbar()
...
                 ->message('JGLOBAL_CONFIRM_DELETE')
                 ->listCheck(true);
         }
 
         if (!$this->isEmptyState && (!$state->get('filter.state') == -2 && $canDo->get('core.delete'))) {
-            $toolbar->confirmButton('delete')
-                ->text('COM_REDIRECT_TOOLBAR_PURGE')
+            $toolbar->confirmButton('delete', 'COM_REDIRECT_TOOLBAR_PURGE')
                 ->message('COM_REDIRECT_CONFIRM_PURGE')
                 ->task('links.purge');
```
Comment (Quy) [inline administrator/components/com_redirect/src/View/Links/HtmlView.php]:
```suggestion
            $toolbar->confirmButton('delete', 'COM_REDIRECT_TOOLBAR_PURGE', 'links.purge')
                ->message('COM_REDIRECT_CONFIRM_PURGE');
```

File: administrator/components/com_installer/src/View/Manage/HtmlView.php [hunk 11]
```diff
@@ -85,35 +85,32 @@ public function display($tpl = null)
...
-                ->text('JTOOLBAR_DISABLE')
+            $toolbar->unpublish('manage.unpublish', 'JTOOLBAR_DISABLE')
                 ->listCheck(true);
             $toolbar->divider();
         }
 
-        $toolbar->standardButton('refresh')
-            ->text('JTOOLBAR_REFRESH_CACHE')
+        $toolbar->standardButton('refresh', 'JTOOLBAR_REFRESH_CACHE')
             ->task('manage.refresh')
```
Comment (Quy) [inline administrator/components/com_installer/src/View/Manage/HtmlView.php]:
```suggestion
        $toolbar->standardButton('refresh', 'JTOOLBAR_REFRESH_CACHE', 'manage.refresh')
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 661f0b17a91b7d97d0f1cb3fbab61e52a4c4b0f8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39537">issues.joomla.org/tracker/joomla-cms/39537</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 661f0b17a91b7d97d0f1cb3fbab61e52a4c4b0f8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39537">issues.joomla.org/tracker/joomla-cms/39537</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39537">issues.joomla.org/tracker/joomla-cms/39537</a>.</sub>

**Comment by obuisard** [comment]
Thank you George @wilsonge for this PR!

**Pull request merged**